distributed/shuffle/_scheduler_extension.py

```diff
@@ -199,7 +293,15 @@ def restart(self, scheduler: Scheduler) -> None:
         self.erred_shuffles.clear()
 
 
-def get_worker_for(output_partition: int, workers: list[str], npartitions: int) -> str:
-    "Get the address of the worker which should hold this output partition number"
+def get_worker_for_range_sharding(
+    output_partition: int, workers: list[str], npartitions: int
+) -> str:
+    """Get address of target worker for this output partition using range sharding"""
     i = len(workers) * output_partition // npartitions
     return workers[i]
+
+
+def get_worker_for_hash_sharding(output_partition: NIndex, workers: list[str]) -> str:
+    """Get address of target worker for this output partition using hash sharding"""
+    i = hash(output_partition) % len(workers)
+    return workers[i]
```

Comment:
Are output_partitions always symmetrical or can they vary significantly?

I assume this hash-mod mapping is a sufficiently uniform distribution but assuming the output chunks are asymmetric we can likely do better with a different mapping that creates a better output distribution.

Out of scope for this PR, just a question

Reply:
_Theoretically_, output partitions can vary significantly, e.g., you could have a chunked array of `((1, 4), (64, 1024))`, which would correspond to partitions of sizes `[(1, 64), (1, 1024), (4, 64), (4, 1024)]`. 

In practice, I'd assume that they are mostly homogeneous. Then again, I'm not an arrays person.